### PR TITLE
Use ST_LineSubstring to speed up spatial index usage in analyser_osmosis_highway_vs_building

### PR DESCRIPTION
**Pull request by frodrigo**
Tested only on one extract.

Original duration: 10min 15s

Using ST_Segmentize of highways and railways. As expected size close to building size are better

100 | 00:10:26 | 626
-- | -- | --
125 | 00:09:22 | 562
150 | 00:07:58 | 478
175 | 00:09:02 | 542
200 | 00:08:01 | 481
250 | 00:09:07 | 547
300 | 00:09:05 | 545
400 | 00:10:02 | 602
600 | 00:10:13 | 613
800 | 00:13:04 | 784

image

So split at 150m.

On this extract ST_Segmentize induce 20% win.

Remove node limit of 100 still using ST_Segmentize, add an extra 20% duration from the initial duration. So keep the limit for now.

### CODE REVIEW

**Comment by jocelynj** [comment]
Just to be clear: does `ST_Segmentize` only adds nodes on ways, to make sure that we don't have a distance greater than N meters?

**Comment by frodrigo** [comment]
Opps. Your are write `ST_Segmentize` is not wat I want. I even surprised it can improve the pref!

I redo this with  `ST_LineSubstring`. The pref improv if good, close to 50% using sub-linestring f 500m.
And good point, remove the clause on large linestring has not effect.

```
100m,06:08
200m,05:41
400m,05:57
500m,05:08
700m,06:44
1000m,07:41
2000m,07:29
5000m,09:25
``

**Comment by jocelynj** [comment]
I really would like to see some tests in this analyser, as the request is more and more complex :) 

If it is not possible to add a test now, we can still merge your change.

**Comment by Famlam** [comment]
If you wish I can add some test cases for the entire analyser to PR #1820, which modifies the same analyser? I have a boring train ride in the afternoon tomorrow :)

If I understand the code well, there's one catch:
splitting the highways in 500m chunks may lose some hits in sql30 and sql31 (as there's more dead points, see image) but simultaneously dropping the 100 nodes restriction may give some new results. I assume this is intentional?
image

File: analysers/analyser_osmosis_highway_vs_building.py [hunk 1]
```diff
@@ -53,8 +56,29 @@
     )) AND
     (NOT tags?'area' OR tags->'area' = 'no') AND
     NOT tags?'area:highway' AND
-    array_length(nodes, 1) <= 100 AND -- Large ways have too big bbox
     ST_NPoints(linestring) > 1
+)
+SELECT
+    id,
+    ST_Transform(linestring_proj_part, 4326) AS linestring,
```
[the remark this answers is not in the document]
Comment (frodrigo) [reply analysers/analyser_osmosis_highway_vs_building.py:63]: No, it convert linestring part back to 4326 for query not working with projected linestring.

[the remark this answers is not in the document]
Comment (frodrigo) [reply analysers/analyser_osmosis_highway_vs_building.py]: Removed to avoid computation error.

**Comment by jocelynj** [comment]
> If you wish I can add some test cases for the entire analyser to PR https://github.com/osm-fr/osmose-backend/pull/1820, which modifies the same analyser?

That would be very nice!

**Comment by Famlam** [comment]
> > If you wish I can add some test cases for the entire analyser to PR #1820, which modifies the same analyser?
> 
> That would be very nice!

Tests are added in #1820

**Pull request closed without merging**